### DISCUSSION TITLE
Pinch and wheel difference across browsers

### DISCUSSION THREAD

**Discussion by breauxna** [pmndrs/use-gesture#518] [Q&A] [answered]
First off, great library! In chrome both pinch and wheel gestures are called for ctrl+wheel and trackpad pinch inputs. In safari only the wheel gesture is called on ctrl+wheel, and only pinch on trackpad. Is this intentional and if so how do you suggest handling an action like zoom when it'll be called twice in chrome (pinch + wheel), but only once in safari (wheel for ctrl+wheel and pinch for trackpad pinch).

**Comment by dbismut** [comment]
This is sort of intentional indeed. When using the trackpad in MacBooks or the Magic Trackpad, Safari is able to determine both zoom and rotation through vendor-specific
`GestureEvents`. But when using those trackpads on Chrome, if you try to zoom, you'll actually trigger `WheelEvents` with the `ctrlKey` flag set to true.

This is not the lib doing anything, this is just the way it is.

However if you want to distinguish wheel from zoom, you can always check state.pinching inside your `onWheel` handler (or check if the `ctrlKey` is false).

**Comment by breauxna** [comment]
Thanks for the quick reply! Ohh ok, I understand. I guess in my case the issue isn't distinguishing wheel from zoom it's just that pinch and wheel both get called in Chrome. So just a small example with some pseudocode of how I'm handling it now

```
const handleWheelStart = useCallback(
    ({ metaKey, ctrlKey, event }) => {
      event.preventDefault();
      event.stopPropagation();

      const isPinch = ctrlKey || metaKey;

      if (isPinch) {
        onZoomStart();
      } else {
        onPanStart();
      }
    },
    [onZoomStart, onPanStart],
  );

const handleWheel = useCallback(
    ({ delta, event, ctrlKey, metaKey, velocity }) => {
      const isPinch = ctrlKey || metaKey;

      if (isPinch) {
        handleZoom()
      } else {
        handlePan()
      }
    },
    [handleZoom, handlePan],
  );

const handleWheelEnd = useCallback(
    ({ metaKey, ctrlKey }) => {
      const isPinch = ctrlKey || metaKey;

      if (isPinch) {
        onZoomEnd();
      } else {
        onPanEnd();
      }
    },
    [onZoomEnd, onPanEnd],
  );

const handlePinchStart = useCallback(() => {
    onZoomStart();
  }, [onZoomStart]);

const handlePinch = useCallback(
    ({ origin, delta }) => {
      handleZoom();
    },
    [handleZoom],
  );

const handlePinchEnd = useCallback(() => {
    onZoomEnd();
  }, [onZoomEnd]);

useGesture(
    {
      onWheelStart: (state) => handleWheelStart(state),
      onWheel: (state) => handleWheel(state),
      onWheelEnd: (state) => handleWheelEnd(state),
      onPinchStart: () => handlePinchStart(),
      onPinch: (state) => handlePinch(state),
      onPinchEnd: () => handlePinchEnd(),
    },
    {
      target: containerRef,
      eventOptions: { passive: false, capture: false },
    },
  );
```

So in Chrome handleZoom will get called twice, once in the wheel handler and once in the pinch handler for either a trackpad pinch or a `ctrl+wheel` action. However, in safari the pinch only gets called on the trackpad and the wheel only gets called on the `ctrl+wheel` so my app will zoom twice as far in chrome on a `ctrl+wheel` then it will in safari. Just trying to figure out if there is a way to handle this better, maybe some gesture option I'm missing or something. Thanks again!

**Comment by breauxna** [comment]
As a simple example I forked this sandbox from one of your examples https://codesandbox.io/s/red-wood-o90q7m?file=/src/App.jsx . If you ctrl+wheel or trackpad pinch on Chrome you will get both the wheel and pinch logs in the console. Could this possibly be a regression? I was looking through some of the old issues and found one that is similar or the same as this, but it was resolved back in 2019. https://github.com/pmndrs/use-gesture/issues/80#issuecomment-521235557

**Comment by breauxna** [comment, marked as answer]
@dbismut It seems like v9 used to only send pinch events in Chrome on `ctrl+wheel` or trackpad pinch in Chrome https://codesandbox.io/s/react-use-gesture-issue-wheel-forked-pmlqou?file=/src/index.js

**Reply by dbismut** [reply]
I'm currently on holidays so not in front of my computer but I don't get why you're handling zoom inside `onWheel`.

It's absolutely **expected** that a wheel event with `ctrlKey: true` would trigger both onPinch and onWheel.

**Reply by breauxna** [reply]
Ohh sorry to bug you then, thanks for responding. I'm handling zoom inside `onWheel` because I have to handle zoom for both trackpad and `ctrl+wheel` in Safari and Chrome. I would use just the pinch handler but that doesn't get triggered on `ctrl+wheel` in Safari, only the wheel handler. I guess I could target the browser from the userAgent in the wheel handler, but I was hoping to avoid that.

**Reply by breauxna** [reply]
```
It's absolutely expected that a wheel event with ctrlKey: true would trigger both onPinch and onWheel.
```

Seems like this isn't the case in Safari. To be fair, I'm not sure if this is a library or browser specific thing though.

**Reply by dbismut** [reply]
Ok I get your issue. At the moment the lib detects whether `WebKitGestureEvents` are available and since React doesn't have a handler to them, the lib only uses them when you pass the `ref` option.

So I guess one easy way to get to what you want is not using `ref`.

That isn't super satisfactory though. 

One other option on your end would be to only use `onPinch` for mobile devices (which you can do by checking [pointerType](https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent/pointerType)) and do the zoom logic in `onWheel` only (in that case you wouldn't benefit from the Magic Trackpad on Safari). 

One final option would be for the lib to still attach wheel listeners on Safari desktop (fairly easy to do). We could have a special option for that.

**Reply by dbismut** [reply]
And yes, to clarify:

- Magic trackpad zoom on Chrome / Firefox triggers the wheel event with the ctrlKey flag set to true.

- Magic trackpad zoom on Safari triggers a GestureEvent.

- usePinch listens to GestureEvents when it can (Safari with ref option), and falls back to TouchEvents / WheelEvents when it can't (when using `{...bind()}`, on Chrome / Firefox).

**Reply by breauxna** [reply]
Ohh ok, that makes sense thanks for the explanation! So in this case I could add a pinch option like `alwaysCapturePinchWheelEvents` or something like that so on a `ctrl+wheel` pinch would still be called? Then I could just handle all my zooming in the pinch handler. I could try to get a PR up for this since you are on vacation depending on exactly how you'd want it done.

**Reply by dbismut** [reply]
I gave it a second thought and you're right. It's better to normalize the behavior of the browsers using the lib and to also track wheel events for Safari no matter whether GestureEvents could be activated. If people want to discard wheel events on Safari, checking `event.type` would suffice. 

I would gladly accept a PR. The fix would be super simple I think, just removing the conditional statement and adding the wheel listener no matter what. 

https://github.com/pmndrs/use-gesture/blob/c0949e078d025fdb3dc80939e173e02e85ec254d/packages/core/src/engines/PinchEngine.ts#L281-L297

I should probably write development guidelines but this package uses `pnpm`, `preconstruct` and `changeset`. If you haven't already you'll have to install `pnpm` globally before installing deps.

Then when you're done with the changes you can run pnpm changeset and the cli will guide you through the steps (this fix should be just a patch). 

Thanks for your help!

**Reply by dbismut** [reply]
Your PR has fixed this and is now released as part of v10.2.17! Thanks for your help @breauxna!
